### PR TITLE
New implementation with API headers, activation formula and flexible column names

### PR DESCRIPTION
**Pull request by akushnarov**

### CODE REVIEW

File: openweather-based-bidding/main.gs [hunk 1]
```diff
@@ -23,8 +23,11 @@ const sheetsApi = new SheetsApi(configSpreadsheetId);
 /**
  * Checks the weather conditions from the Open Weather API and adjusts the
  * DV360 entities status (e.g. IO switched on/off) with DV360 API.
+ * 
+ * @param {bool} onlyCheckAPI Set to true if you want only check API (no DV360 sync)
```
Comment (paranerd) [inline openweather-based-bidding/main.gs:27]: `if you want _to_ only check _the_ API`

File: openweather-based-bidding/main.gs [hunk 2]
```diff
@@ -45,6 +48,19 @@ function monitorWeatherAndSyncWithDV360() {
...
 
+    // Check if we already processed this item
+    const currentDateTime = new Date();
+    const lastUpdated = new Date(
+      row[ config.getHeaderIndex('col-last-updated') ]
+    );
+    
+    const diff = (currentDateTime - lastUpdated) / 1000 / 60 / 60;
+    const howOften = parseInt(config.get('how-often-to-check-in-hours'));
+    if ( !onlyCheckAPI && howOften && diff < howOften) {
```
Comment (paranerd) [inline openweather-based-bidding/main.gs:59]: Let's be consistent:
`if ( !onlyCheckAPI...` vs.
`if (! onlyCheckAPI...` vs.
`if (!sheetsApi.write...`

I'd prefer the last one.

Also, according to my comment above:
`hoursBetweenUpdates`

File: openweather-based-bidding/main.gs [hunk 3]
```diff
@@ -45,6 +48,19 @@ function monitorWeatherAndSyncWithDV360() {
...
     const iPlus1 = i + 1;
 
+    // Check if we already processed this item
+    const currentDateTime = new Date();
+    const lastUpdated = new Date(
+      row[ config.getHeaderIndex('col-last-updated') ]
+    );
+    
+    const diff = (currentDateTime - lastUpdated) / 1000 / 60 / 60;
+    const howOften = parseInt(config.get('how-often-to-check-in-hours'));
```
Comment (paranerd) [inline openweather-based-bidding/main.gs:58]: According to my comment above:
`const hoursBetweenUpdates = parseInt(config.get('hours-between-updates'));`

File: openweather-based-bidding/classes/sheets.gs [hunk 4]
```diff
@@ -52,10 +64,69 @@ class SheetsApi {
...
+
+  /**
+   * Fetches data from one cell from the sheet
+   *
+   * @param {string} row Row number
+   * @param {string} col Column number
+   *
+   * @return {Object}
+   */
+  getOne(row, col) {
```
Comment (paranerd) [inline openweather-based-bidding/classes/sheets.gs:88]: Maybe call it `getCell()` to be more clear? Or even `getCellValue()`?

File: openweather-based-bidding/classes/config.gs [hunk 5]
```diff
@@ -58,23 +58,38 @@ class Config {
...
+    this.config.headers = headers;
+  }
+
+  /**
+   * Returns the index of the header entity
+   * 
+   * @param idx Header notation (from `this.config`) we are looking for
+   * @return {integer} Index, if not exists then -1.
+   */
+  getHeaderIndex(idx, plusOne) {
```
Comment (paranerd) [inline openweather-based-bidding/classes/config.gs:97]: `plusOne` is missing in the doc string.
I suggest calling it `zeroBased` and if `!zeroBased` -> add 1 (or let the caller deal with that entirely)

File: openweather-based-bidding/main.gs [hunk 6]
```diff
@@ -26,97 +26,66 @@ const sheetsApi = new SheetsApi(configSpreadsheetId);
...
+          advertiserId = parseInt(
+            row[config.getHeaderIndex('col-advertiser-id')]
+          ),
+          lat = parseFloat(row[config.getHeaderIndex('col-lat')]),
+          lon = parseFloat(row[config.getHeaderIndex('col-lon')]);
+
+    // Get weather conditions
+    const allWeather = weather.getAll(lat, lon);
+
+    // Extract all weather variables and write their values to the spreadsheet
```
Comment (paranerd) [inline openweather-based-bidding/main.gs:77]: We're not writing to the spreadsheet at this point.
`// Extract weather conditions` may be sufficient?

File: openweather-based-bidding/main.gs [hunk 7]
```diff
@@ -26,97 +26,66 @@ const sheetsApi = new SheetsApi(configSpreadsheetId);
...
 function monitorWeatherAndSyncWithDV360() {
   Logger.log('[START] monitorLineItemChangesAndSyncWithDV360');
+
   // Get items from Sheet
-  const data = sheetsApi.get(configSpreadsheetName);
-  const rows = data['values'];
+  const rows = sheetsApi.get(configSpreadsheetName);
+
+  // Process sheet headers
+  const apiHeaders = Utils.getApiHeaders(rows[0]);
```
Comment (paranerd) [inline openweather-based-bidding/main.gs:37]: This confused me quite a bit before I realized that `setHeaders()` in the next line does not process the `apiHeaders`^^
Could we move this down to where `apiHeaders` is actually used?

**Review by paranerd — CHANGES_REQUESTED** [review]

File: openweather-based-bidding/classes/utils.gs [hunk 8]
```diff
@@ -41,4 +41,44 @@ class Utils {
...
+  static getValueFromJSON(path, json) {
+    let tmpJson  = json, 
+        val       = null;
+    
+    for (const part of path.split('.')) {
+      console.log('part', part);
+      let tmpVal;
+      const intVal = parseInt(part);
+      if (intVal && intVal in tmpJson) {
+        console.log(61);
```
[the remark this answers is not in the document]
Comment (akushnarov) [reply openweather-based-bidding/classes/utils.gs]: This was one of the intermediate commits, this `console.log()` was removed by a later commit and I don't see it here https://github.com/google/dv360-api/pull/3/files, so should be fixed already.

File: openweather-based-bidding/classes/open-weather.gs [hunk 9]
```diff
@@ -29,17 +29,23 @@ class OpenWeather {
...
-   * Get current and forecast weather data based on geo location
+   * Get weather JSON from API based on geo location
    *
    * @param {number} lat
    * @param {number} lon
    *
    * @returns {Object}
    */
-  getCurrent(lat, lon) {
+  getAll(lat, lon) {
```
[the remark this answers is not in the document]
Comment (akushnarov) [reply openweather-based-bidding/classes/open-weather.gs:43]: Good point!

File: openweather-based-bidding/main.gs [hunk 10]
```diff
@@ -64,9 +64,11 @@ function monitorWeatherAndSyncWithDV360() {
         .getValueFromJSON(apiHeader, allWeather);
     }
 
+    row[ config.getHeaderIndex('col-last-updated') ] = new Date().toISOString();
+
     // Save weather conditions back to Sheet
     if (!sheetsApi.write([row], configSpreadsheetName + '!A' + iPlus1)) {
-      Logger.log('An error occurred, retrying in 30s');
+      Logger.log('(1) An error occurred, retrying in 30s');
```
[the remark this answers is not in the document]
Comment (akushnarov) [reply openweather-based-bidding/main.gs:89]: That's true, will fix

File: openweather-based-bidding/main.gs [hunk 11]
```diff
@@ -78,14 +80,25 @@ function monitorWeatherAndSyncWithDV360() {
...
-      dv360.switchIOStatus(advertiserId, insertionOrderId, activate);
+    try {
+      // Switch Status according to the activation formula value
+      if (!isNaN(lineItemId) && lineItemId > 0) {
+        dv360.switchLIStatus(advertiserId, lineItemId, activate);
+      } else if (!isNaN(insertionOrderId) && insertionOrderId > 0) {
+        dv360.switchIOStatus(advertiserId, insertionOrderId, activate);
+      }
+    } catch (e) {
+      Logger.log('(2) An error occurred, retrying in 30s');
```
[the remark this answers is not in the document]
Comment (akushnarov) [reply openweather-based-bidding/main.gs]: will fix

File: openweather-based-bidding/main.gs [hunk 12]
```diff
@@ -78,14 +80,25 @@ function monitorWeatherAndSyncWithDV360() {
...
+      
+      // Decrement `i` so that it ends up the same in the next for-loop iteration
+      i--;
     }
     
-    // +TODO: Write logs!
+    // Logging
+    row[ config.getHeaderIndex('col-formula') ] = activate;
+    row.push('[ROW DATA]');
+    Logger.log(row.join(','));
```
[the remark this answers is not in the document]
Comment (akushnarov) [reply openweather-based-bidding/main.gs]: The main idea here was to 
- save all the table values in CSV format (for the future additional analysis, e.g. when and which LI was on/off)
- [ROW DATA] is added as a label, so user can filter the logs by this label and download them as CSV (if this label was on the separate logs line, additional data processing would be needed)

File: openweather-based-bidding/tests/Utils.test.js [hunk 13]
```diff
@@ -31,7 +31,7 @@ test('encodeParameters', () => {
 });
 
 test('getValueFromJSON', () => {
-    const json = {
+    let json = {
```
[the remark this answers is not in the document]
Comment (akushnarov) [reply openweather-based-bidding/tests/Utils.test.js]: Usually I'm trying to re-use existing variables to save memory (not sure this is applicable here).

File: openweather-based-bidding/tests/Utils.test.js [hunk 14]
```diff
@@ -56,6 +56,26 @@ test('getValueFromJSON', () => {
...
+            },
+        ], 
+        'bar': {
+            'bar1': 'foo1',
+        }
+    };
+
+    expect(
+        Utils.getValueFromJSON('daily.0.weather.id', json)
+    ).toBe(123);
```
[the remark this answers is not in the document]
Comment (akushnarov) [reply openweather-based-bidding/tests/Utils.test.js:78]: Yes, removing

File: openweather-based-bidding/classes/config.gs [hunk 15]
```diff
@@ -68,6 +68,14 @@ class Config {
...
       // Spreadsheet headers
       'headers': [],
+
+      /**
+       * *Optional*, how often do you want to check data from the API
+       * and switch the marketing entity (IO or LI). By default this is
+       * 12 hours (two times per day). Leave empty if you don't want to
+       * check this option.
+       */
+      'how-often-to-check-in-hours': 12,
```
[the remark this answers is not in the document]
Comment (akushnarov) [reply openweather-based-bidding/classes/config.gs:78]: Yes, thanks. Will fix

File: openweather-based-bidding/main.gs (hunk 1 above)
Comment (akushnarov) [reply openweather-based-bidding/main.gs:27]: Thanks 👍🏿

File: openweather-based-bidding/main.gs (hunk 2 above)
Comment (akushnarov) [reply openweather-based-bidding/main.gs:59]: Totally agree, will review all if statements. Also would be nice to agree on the coding guidelines.

File: openweather-based-bidding/main.gs (hunk 3 above)
Comment (akushnarov) [reply openweather-based-bidding/main.gs:58]: Will change

File: openweather-based-bidding/classes/sheets.gs (hunk 4 above)
Comment (akushnarov) [reply openweather-based-bidding/classes/sheets.gs:88]: 👍🏿

File: openweather-based-bidding/classes/config.gs (hunk 5 above)
Comment (akushnarov) [reply openweather-based-bidding/classes/config.gs:97]: Good point. I decided to rename it to the "indexShift" and slightly refactored the function

File: openweather-based-bidding/main.gs (hunk 6 above)
Comment (akushnarov) [reply openweather-based-bidding/main.gs:77]: Good one, thanks

File: openweather-based-bidding/main.gs (hunk 7 above)
Comment (akushnarov) [reply openweather-based-bidding/main.gs:37]: I'm did a refactoring and moved the header functionality to the `config` module. Please let me know if this is more readable
```
  config.setHeaders(rows[0]);
  const apiHeaders = config.getApiHeaders();
```

**Comment by akushnarov** [comment]
@paranerd please see the changes from the last code review https://github.com/google/dv360-api/pull/3/commits/c6c59dd52ba34f11467f648d60f2c7bd7ce70082

**Review by paranerd — APPROVED** [review]

**Comment by akushnarov** [comment]
Makes sense (also this feature was used only once), thanks. Fixed with the commit https://github.com/google/dv360-api/pull/3/commits/d2fc22a163d2fa0284c61031e51cbc2017b09221.

**Review by paranerd — APPROVED** [review]

**Pull request merged**